read_buffer/src/table.rs

```diff
@@ -440,88 +420,92 @@ struct MetaData {
-                .column_ranges()
-                .iter()
-                .map(|(k, v)| (k.to_string(), (v.0.clone(), v.1.clone())))
-                .collect(),
-            column_types: rg
-                .column_logical_types()
-                .iter()
-                .map(|(k, v)| (k.clone(), *v))
-                .collect(),
-            column_names: rg.column_ranges().keys().cloned().collect(),
-            time_range: Some(rg.time_range()),
+            size: meta.size,
+            rows: meta.rows as u64,
+            columns: meta.columns.clone(),
+            column_names: meta.columns.keys().cloned().collect(),
+            time_range: Some(meta.time_range),
         }
     }
 
     // Extract schema information for a set of columns. If a column name does
     // not exist within the `Table` schema it is ignored and not present within
     // the resulting schema information.
-    fn schema_for_column_names(&self, names: &[ColumnName<'_>]) -> Vec<(String, LogicalDataType)> {
+    fn schema_for_column_names(
+        &self,
+        names: &[ColumnName<'_>],
+    ) -> Vec<(ColumnType, LogicalDataType)> {
         names
             .iter()
-            .filter_map(|&name| match self.column_types.get_key_value(name) {
-                Some((name, data_type)) => Some((name.clone(), *data_type)),
+            .filter_map(|&name| match self.columns.get(name) {
+                Some(schema) => Some((schema.typ.clone(), schema.logical_data_type)),
                 None => None,
             })
             .collect::<Vec<_>>()
     }
 
+    // As `schema_for_column_names` but for all columns in the table.
+    fn schema_for_all_columns(&self) -> Vec<(ColumnType, LogicalDataType)> {
+        self.columns
+            .iter()
+            .map(|(name, schema)| (schema.typ.clone(), schema.logical_data_type))
+            .collect::<Vec<_>>()
+    }
+
     // As `schema_for_column_names` but also embeds the provided aggregate type.
     fn schema_for_aggregate_column_names(
         &self,
         names: &[(ColumnName<'_>, AggregateType)],
-    ) -> Vec<(String, AggregateType, LogicalDataType)> {
+    ) -> Vec<(ColumnType, AggregateType, LogicalDataType)> {
         names
             .iter()
-            .filter_map(
-                |(name, agg_type)| match self.column_types.get_key_value(*name) {
-                    Some((name, data_type)) => Some((name.clone(), *agg_type, *data_type)),
-                    None => None,
-                },
-            )
-            .collect::<Vec<_>>()
-    }
-
-    // Extract all schema information for the `Table`.
-    fn schema(&self) -> Vec<(String, LogicalDataType)> {
-        self.column_types
-            .iter()
-            .map(|(k, v)| (k.clone(), *v))
+            .filter_map(|(name, agg_type)| match self.columns.get(*name) {
+                Some(schema) => Some((schema.typ.clone(), *agg_type, schema.logical_data_type)),
+                None => None,
+            })
```

Comment:
You can avoid the `None => None` thing if you want by calling `map` on the Option returned from `get`:

```suggestion
            .filter_map(|&name| {
                self.columns.get(name)
                    .map(|schema| (schema.typ.clone(), schema.logical_data_type))
            })

```

Reply:
I keep forgetting this! I need `agg_type` though I fixed this in e662a519